src/rpc/rawtransaction.cpp

```diff
@@ -922,67 +924,77 @@ static RPCHelpMan testmempoolaccept()
+    const UniValue raw_transactions = request.params[0].get_array();
     const CFeeRate max_raw_tx_fee_rate = request.params[1].isNull() ?
                                              DEFAULT_MAX_RAW_TX_FEE_RATE :
                                              CFeeRate(AmountFromValue(request.params[1]));
 
     CTxMemPool& mempool = EnsureMemPool(request.context);
-    int64_t virtual_size = GetVirtualTransactionSize(*tx);
-    CAmount max_raw_tx_fee = max_raw_tx_fee_rate.GetFee(virtual_size);
-
-    UniValue result(UniValue::VARR);
-    UniValue result_0(UniValue::VOBJ);
-    result_0.pushKV("txid", tx->GetHash().GetHex());
-    result_0.pushKV("wtxid", tx->GetWitnessHash().GetHex());
+    std::vector<CTransactionRef> txns;
 
-    TxValidationState state;
-    bool test_accept_res;
-    CAmount fee{0};
-    {
-        LOCK(cs_main);
-        test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx),
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, /* test_accept */ true, &fee);
+    for (unsigned int i = 0; i < raw_transactions.size(); ++i) {
+        CMutableTransaction mtx;
+        if (!DecodeHexTx(mtx, raw_transactions[i].get_str())) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
+        }
+        txns.emplace_back(MakeTransactionRef(std::move(mtx)));
     }
 
-    // Check that fee does not exceed maximum fee
-    if (test_accept_res && max_raw_tx_fee && fee > max_raw_tx_fee) {
-        result_0.pushKV("allowed", false);
-        result_0.pushKV("reject-reason", "max-fee-exceeded");
-        result.push_back(std::move(result_0));
-        return result;
+    std::vector<MempoolAcceptResult> validation_results;
+    if (txns.size() == 1) {
+        validation_results.emplace_back(WITH_LOCK(cs_main,
+            return AcceptToMemoryPool(mempool, txns[0], false /* bypass_limits */, true /* test_accept */)));
+    } else {
+        validation_results = WITH_LOCK(cs_main, return ProcessNewPackage(mempool, txns, true));
     }
-    result_0.pushKV("allowed", test_accept_res);
 
-    // Only return the fee and vsize if the transaction would pass ATMP.
-    // These can be used to calculate the feerate.
-    if (test_accept_res) {
-        result_0.pushKV("vsize", virtual_size);
-        UniValue fees(UniValue::VOBJ);
-        fees.pushKV("base", ValueFromAmount(fee));
-        result_0.pushKV("fees", fees);
-    } else {
-        if (state.IsInvalid()) {
-            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
-                result_0.pushKV("reject-reason", "missing-inputs");
+    auto tx_accepted = [](MempoolAcceptResult& res) { return res.m_accepted; };
+    const bool all_valid = std::all_of(validation_results.begin(), validation_results.end(), tx_accepted);
+    // ProcessNewPackage should return a MempoolAcceptResult per transaction
+    // or exactly 1 for the first error that occurs.
+    CHECK_NONFATAL(all_valid || validation_results.size() == 1);
```

Comment:
I'm not very familiar with `CHECK_NONFATAL` but see it throws an error if either condition here evaluates to false. 

If `validation_results.size() == 1` that means one of the transactions failed validation, so we want to throw an error. Should this line be changed to `CHECK_NONFATAL(all_valid || validation_results.size() != 1);`?

Reply:
`CHECK_NONFATAL` (documented in developer notes [here](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#assertions-and-checks)) helps with internal logic bugs. Here, the bug would be a testmempoolaccept thing only (and therefore not be fatal enough to crash the node), but we'd definitely want to address it.

Here, we are merely making sure that, if there were any errors, the length of `validation_results` should be 1. It's okay for a transaction to be invalid so we don't want to throw an error for that. This check is just to make sure that `ProcessNewPackage()` is adhering to the API we expect. I hope this helps!

Reply:
Ah! Thanks for linking to that description 👍  [This was](https://doxygen.bitcoincore.org/check_8h.html#a46a3e27097aa5e94bbf62075bad7016f) the only one I had been working off of so the link you provided is super helpful for understanding the purpose of CHECK_NONFATAL. 

I was getting hung up on the `||`, or part of the evaluation. I think I understand it now, an error will only be thrown if both `all_valid` and `validation_results.size() == 1` evaluate to false, reason being that these two statements are incompatible (if there is an invalid tx, `validation_results.size()` can't be anything other than 1). Thanks for the clarification! Please mark this guy as resolved when you get a chance! (I don't believe I have permission to)